test/Npgsql.Tests/ConnectionTests.cs

```diff
@@ -1559,6 +1559,102 @@ public async Task TimeoutDuringAuthentication()
                     .With.InnerException.TypeOf<TimeoutException>());
         }
 
+        [Test]
+        public async Task Physical_open_callback_sync()
+        {
+            await using var defaultConn = await OpenConnectionAsync();
+            await using var _ = await CreateTempTable(defaultConn, "ID INTEGER", out var table);
+
+            using var __ = CreateTempPool(ConnectionString, out var connectionString);
+            using var conn = new NpgsqlConnection(connectionString);
+            conn.PhysicalOpenCallback = connector =>
+            {
+                using var cmd = connector.CreateCommand($"INSERT INTO \"{table}\" VALUES(1)");
+                cmd.ExecuteNonQuery();
+                return new ValueTask();
+            };
+
+            conn.Open();
+
+            var rowsCount = (long)(await conn.ExecuteScalarAsync($"SELECT COUNT(*) FROM \"{table}\""))!;
+            Assert.AreEqual(1, rowsCount);
+        }
+
+        [Test]
+        public async Task Physical_open_callback_async()
+        {
+            await using var defaultConn = await OpenConnectionAsync();
+            await using var _ = await CreateTempTable(defaultConn, "ID INTEGER", out var table);
+
+            using var __ = CreateTempPool(ConnectionString, out var connectionString);
+            await using var conn = new NpgsqlConnection(connectionString);
+            conn.PhysicalOpenCallback = async connector =>
+            {
+                using var cmd = connector.CreateCommand($"INSERT INTO \"{table}\" VALUES(1)");
+                await cmd.ExecuteNonQueryAsync();
+            };
+
+            await conn.OpenAsync();
+
+            var rowsCount = (long)(await conn.ExecuteScalarAsync($"SELECT COUNT(*) FROM \"{table}\""))!;
+            Assert.AreEqual(1, rowsCount);
+        }
+
+        [Test]
+        public async Task Physical_open_callback_async_over_sync()
+        {
+            using var _ = CreateTempPool(ConnectionString, out var connectionString);
+            await using var conn = new NpgsqlConnection(connectionString);
+            conn.PhysicalOpenCallback = async connector =>
+            {
+                await Task.Delay(50);
```

Comment:
This could be flaky...

```suggestion
                await Task.Yield();
```

Reply:
Was thinking about this, but wouldn't `Task.Yield` be flaky as well?

Reply:
How so? I think Task.Yield is guaranteed to return an incomplete Task - otherwise it wouldn't always do its one job of yielding, no? Whereas the Task returned by Task.Delay may be complete by the time its state is checked, no?

Reply:
> How so? I think Task.Yield is guaranteed to return an incomplete Task - otherwise it wouldn't always do its one job of yielding, no?

It does return an incomplete Task, but is it still incomplete by the time we check it? Let's try with `Yield` and see what happens.

Reply:
Yield is actually returning an awaiter which always reports false when calling IsCompleted. This forces any async machinery to unwind the stack and resume via the continuation attached to the YieldAwaiter (which is triggered immediately).